vignettes/BERT_basics.Rmd

```diff
@@ -0,0 +1,165 @@
+
+## The output of BERT
+
+### Embeddings 
+
+One way of thinking about BERT is as a machine for producing context-dependent
+[embeddings](https://en.wikipedia.org/wiki/Word_embedding). Here, an "embedding"
+is a vector [^vector] that gets associated with each token of the input. Useful
+embeddings will have a number of special properties. For example, tokens with
+similar meanings will have embeddings that are "nearby" in the embedding space.
+
+Static embeddings, such as [word2vec](https://en.wikipedia.org/wiki/Word2vec),
+have been around for several years. However, they typically have been unable to
+distinguish between [homographs](https://en.wikipedia.org/wiki/Homograph), such
+as "train [teach]" and "train [locomotive]". More generally, such embeddings are
+insensitive to word order, sentence structure, and other contextual cues.
+
+[^vector]: Think, "point in a high-dimensional space."
+
+### Context-dependent embeddings
+
+In contrast, BERT's output can be understood as embedding vectors that *are*
+appropriately sensitive to the context in which each word is used. Not only does
+this make it possible to give homographs their own embeddings, it also allows
+more subtle differences in meaning and usage to be picked up.
+
+BERT outputs an embedding vector for each input token, including the special
+tokens "[CLS]" and "[SEP]". The vector for [CLS] can be thought of as "pure
+context"; it's the embedding of a token that has no intrinsic meaning, but is
+still sensitive to the context around it.[^CLS]
+
+BERT has a layered structure (see next section), and output embedding vectors
+can be obtained at each layer.
+
+[^CLS]: The interpretation of [CLS] is a bit more nuanced than this simple
+explanation implies. See [this
+discussion](https://github.com/google-research/bert/issues/196) for more
+details.
+
+## The insides of BERT
+
+Context-dependence is achieved through the "attention" mechanism. In very
+cartoony terms, "attention" provides a way for each token to "choose" (based on
+training) which of its surrounding tokens to be most influenced by. BERT
+consists of multiple sequential layers of attention, with multiple "heads" per
+layer. Each head may split its attention across any of the input tokens
+(including itself). It may be helpful to picture each token processed by BERT as
+a many-headed beast, able to attend at each moment to any or all of its
+neighbors, and modify itself slightly in the next moment based on what it sees.
+
+The amount of attention paid by each token, to each token, in each layer and
+head, can be represented by a weight that is normalized to one for each
+"attender". While these weights are not part of the formal output of BERT, it
+can be instructive to study them to better understand what aspects of language
+BERT models well.
+
+RBERT makes it easy (using the `extract_features` function) to obtain both the
+attention weights and the token embeddings at each layer of a BERT model.
+
```

Comment:
```suggestion

```
Delete the extra carriage return at the end. I don't know why, but RStudio even has a setting to make sure there's a single carriage return at the end of files, 'cuz apparently some things care.

Reply:
BTW, this is a realllllly good intro to BERT! Great work!